### DISCUSSION TITLE
Is it possible to add route query params without refreshing the page ?

### DISCUSSION THREAD

**Discussion by snake-py** [vikejs/vike#1111] [Discussion]
Hello again,

I want to control some basic app states via route query params:

For instance:

```
any-route?modal=name-of-modal-1&step=step-name
```

The above query param tells my application that it should open the modal and navigate to the correct step. However, currently for that I am using some self made hook, which basically uses `navigatè` from vps. So the page refresh basically makes the rerender possible. In the past I might have used a set up where I keep the state in JS via redux and then pushed to borwser history.

So the question is it possible to update query params with some VPS function which will not reload the page, but to which I could listen to via a useEffect? 

I am only using Client Side routing

(I asked in the weekly, but my boss is sick this week - will try next week.)

I have a bit of a hacky solution:

```ts
  useEffect(() => {
    if (typeof window !== 'undefined') {
      window.history.pushState = new Proxy(window.history.pushState, {
        apply: (target, thisArg, argArray) => {
          console.log('pushState', argArray[2]);
          const url = new URL(argArray[2]);
          const urlParams = new URLSearchParams(url.search);
          const myParam = urlParams.get('myNewUrlQuery');
          console.log('pushState', myParam);
          setQueryParam(myParam ?? 'none');
          //@ts-ignore
          return target.apply(thisArg, argArray);
        }
      });
    }
  }, []);
```

**Comment by brillout** [comment]
> update query params with some VPS function which will not reload the page

To be clear: you mean that doesn't trigger a *re-render*, correct? Because if you `navigate()` to the same page then it shouldn't trigger a page reload, only a page re-render (if it does then I consider it to be a bug).

Why doesn't `navigate('any-route?modal=name-of-modal-1&step=step-name')` work out for you? The state is then available at `pageContext.urlParsed.search` &mdash; no need for redux & co.

> (I asked in the weekly, but my boss is sick this week - will try next week.)

🙏 Keep me updated!

**Reply by snake-py** [reply]
> To be clear: you mean that doesn't trigger a re-render, correct? Because if you navigate() to the same page then it shouldn't trigger a page reload, only a page re-render (if it does then I consider it to be a bug).

Yes navigate of course triggers the page to rerender. But I want a way to only store some state in the url like a query param, but I don't want that the whole page refreshes. I want to control myself what should rerender and what should not rerender, if the query params change. 

This is useful for many ways. In my current case I would store the state of the open modal in the query param, so I can open the modal as deep link via email or that can use the back and forward functionality of the browser.

Another example where I would want to use this is a search where I only want to refresh the part of the page where I show filtered items, but nothing else. I could now store this as state but then I would loose the ability of using back and forward buttons. So basically means I would need to update the state myself and then push to history: 

This is how it is right now
```tsx
import { useState } from 'react';
import { navigate } from 'vite-plugin-ssr/client/router';

const products = [
  {
    id: 1,
    name: 'Product 1',
    price: 100,
    description: 'This is product 1'
  },
  {
    id: 2,
    name: 'Product 2',
    price: 200,
    description: 'This is product 2'
  },
  {
    id: 3,
    name: 'Product 3',
    price: 300,
    description: 'This is product 3'
  },
  {
    id: 4,
    name: 'TV',
    price: 300,
    description: 'This is product 3'
  }
];

function Page(pageContext) {
  const [filteredProducts, setFilteredProducts] = useState(products);
  const [filter, setFilter] = useState(pageContext.urlParsed.search.get('filter') || '');
  const onSearch = () => {
    navigate(`/products?filter=${filter}`);
  };
  return (
    <div>
      <h1>Products</h1>
      <input
        type="text"
        onChange={(e) => setFilter(e.target.value)}
      />
      <button onClick={() => onSearch()}>Search</button>

      <ul>
        {products
          .filter((product) => product.description.includes(filter))
          .map((product) => (
            <li key={product.id}>
              <a href={`/products/${product.id}`}>{product.name}</a>
            </li>
          ))}
      </ul>
    </div>
  );
}
```

But what if I would only want to refresh the inside of the `ul`? Then I could not use `navigate` I would need to do something where I manually push the state to two positions the query param (like I showed in my original post), and a normal state. 

The questions now is: 
1.  Is there a VPS way of doing only partial rerenders?
2. Is there a way of appending query params to the url, without a page reload, but which would reflect in the pageProps or some other state?

**Reply by snake-py** [reply]
@brillout 

> 🙏 Keep me updated!

I spoke with them today and boss said, they are looking into it. They want to have a generalized system on how to support OS project which we use. I know them for some time now. If they say "this", it means it will take some time but they usually do something after they have thought on how to implement a proper internal system for this. It might take up to 1 to 3 month until they have a solution.

**Reply by brillout** [reply]
Thanks, I appreciate it a lot. This makes a *big* difference. (Btw. feel free to let me know if you happen to be in Munich and we can go for dinner / some drinks if you want.)

Exciting!

**Reply by brillout** [reply]
(I didn't forget and I'll reply to the `navigate()` discussion later today.)

**Reply by snake-py** [reply]
> Btw. feel free to let me know if you happen to be in Munich and we can go for dinner / some drinks if you want.

This would be too cool, I saw the other day that you are in Munich. I currently live in Italy, but will move back to Germany soon. If I come across Munich I'll let you know. Would be quite exciting tbh.

**Reply by brillout** [reply]
> Yes navigate of course triggers the page to rerender. But I want a way to only store some state in the url like a query param, but I don't want that the whole page refreshes. I want to control myself what should rerender and what should not rerender, if the query params change.

The page doesn't refresh nor reloads, it re-renders. I'm being pedantic here because it makes a big difference. A page reload means that every state of the page is lost. Whereas, when re-rendering a page, React is smart enough to preserve state. This means that, if you `navigate('/same-page?some=query')`, then React will, for example, preserve the state of your search results (as long as you don't render a different React element). By a matter of fact, if you'd render the exact same React elements, then the DOM won't even change at all.

> This is useful for many ways. In my current case I would store the state of the open modal in the query param, so I can open the modal as deep link via email or that can use the back and forward functionality of the browser.

I don't see a problem with that use case when using `navigate('/same-page?some=query')` with `pageContext.urlParsed.search`.

> Another example where I would want to use this is a search where I only want to refresh the part of the page where I show filtered items, but nothing else.

In general, React encourages complete re-renders (that's part of React's philosophy). As for performance, the idea is to memoize components when re-rendering is getting slow. FYI that's also what you'd do with Redux: every time the global store changes, the whole page is re-rendered.

> I could now store this as state but then I would loose the ability of using back and forward buttons.

I believe this also shouldn't be a problem.

So I'd suggest going for the `navigate()` solution and, if you hit performance issues, then use memoization.

**Reply by brillout** [reply]
> This would be too cool, I saw the other day that you are in Munich. I currently live in Italy, but will move back to Germany soon. If I come across Munich I'll let you know. Would be quite exciting tbh.

Looking forward to it 🍻 It's always nice to meet users :)

**Reply by brillout** [reply]
> * Is there a VPS way of doing only partial rerenders?
> * Is there a way of appending query params to the url, without a page reload, but which would reflect in the pageProps or some other state?

Vite-pugin-ssr in itself doesn't trigger the re-render, it merely triggers your `render()` hook. It is your `render()` hook that re-renders the page. So, in principle, you have full control over what happens when, for example, merely the query params changes: you can implement what(ever) you want. But I'd recommend against it as you'll complexify your React integration, unless you want to experiment and have fun. FYI [vike-react](https://github.com/vikejs/vike-react/tree/main/vike-react) keeps things simple and doesn't include such functionality.

**Reply by snake-py** [reply]
> The page doesn't refresh nor reloads, it re-renders. I'm being pedantic here because it makes a big difference. A page reload means that every state of the page is lost. Whereas, when re-rendering a page, React is smart enough to preserve state. This means that, if you navigate('/same-page?some=query'), then React will, for example, preserve the state of your search results (as long as you don't render a different React element). By a matter of fact, if you'd render the exact same React elements, then the DOM won't even change at all.

This is something, I need to try out. Because now it seems like everything re-renders and all state is lost. Maybe we did somethign wrong with our set up? I will investigate this tomorrow and drop you a message.  

> I don't see a problem with that use case when using navigate('/same-page?some=query') with pageContext.urlParsed.search

Then I probably do something wrong, as I said all state seems to be lost, if we use navigate like this. :thinking: 

Thank you for the clarifications!

**Reply by brillout** [reply]
Yea something's wrong if you lose state even though you render the same React elements. I'd inject a Counter at the root and see whether its state is preserved.

**Reply by snake-py** [reply]
@brillout Yes you were right! We have one provider which seems to cause a total loss of state. Need to investigate further why it happens. But thank you for bringing me on the right track again!
